File: frontend/types/expense.ts
export interface Expense {
  id: string;
  amount_rupees: string;  // ALWAYS a string like "199.50" — never a number
  category: string;
  description: string;
  date: string;
  created_at: string;
}

export interface CreateExpenseInput {
  amount_rupees: number;
  category: string;
  description: string;
  date: string;
  idempotency_key: string;
}

export interface User {
  id: string;
  email: string;
  created_at: string;
}

File: backend/app/schemas/expense.py
from pydantic import BaseModel, field_validator
from datetime import date, datetime
from typing import Optional
from app.models.expense import Expense


class CreateExpenseRequest(BaseModel):
    amount_rupees: float
    category: str
    description: str
    date: date
    idempotency_key: str

    @field_validator("amount_rupees")
    @classmethod
    def validate_amount_rupees(cls, v):
        if v <= 0:
            raise ValueError("Amount must be greater than 0")
        return v

    @field_validator("category", "description")
    @classmethod
    def strip_strings(cls, v):
        return v.strip()


class ExpenseResponse(BaseModel):
    id: str
    amount_rupees: str
    category: str
    description: str
    date: date
    created_at: datetime

    @classmethod
    def from_expense(cls, expense: Expense) -> "ExpenseResponse":
        return cls(
            id=str(expense.id),
            amount_rupees=f"{expense.amount_paise / 100:.2f}",
            category=expense.category,
            description=expense.description,
            date=expense.date,
            created_at=expense.created_at,
        )